backends/cxxrtl/cxxrtl_backend.cc

```diff
@@ -864,7 +864,7 @@ struct CxxrtlWorker {
 		if (!module->has_attribute(ID(cxxrtl_template)))
 			return {};
 
-		if (module->attributes.at(ID(cxxrtl_template)).flags != RTLIL::CONST_FLAG_STRING)
```

Comment:
Here we are giving the flag a semantical significance, I think we may want to remove this check

Reply:
I'm going to reject this change.

Reply:
As far as I'm concerned the `STRING` flag has semantic significance and CXXRTL treats it as such in many places (e.g. the types are going to be different and this will change the ABI). I don't really care if RTLIL claims there's no significance to it, fix RTLIL then.

Reply:
With this PR we are removing the semantical significance of `STRING`, instead it serves to signify the backing used, which can change merely by using the wrong accessor on the Const object. We may not go through with it, but that's the plan. As an alternative we can add an extra flag which would signify the backing.

FWIW the `==` and `!=` operators on Const disregard the flags as of today, so we have an inconsistence either way.

Reply:
CXXRTL relies on the semantic significance so I'll not accept this PR until there is a way forward for it.

Reply:
> With this PR we are removing the semantical significance of STRING, instead it serves to signify the backing used, which can change merely by using the wrong accessor on the Const object.

Looking at the code change that may not be true (but it's what I thought we were doing). So the backing seems decided by the state of the `std::variant<>` field, in which case I need to take back what I said about the `STRING` flag.

We still have a somewhat inconsistent state with the comparison operators ignoring the flags, but that may be exposed through to the user in some places and I am not sure we can change it.

Reply:
Note also that this conditional isn't the only (nor the most important) place where this is done; the attribute set dumping code is the primary one. In particular the results of this transformation are used for ChipFlow tools and we rely on it being done correctly.

Reply:
> We still have a somewhat inconsistent state with the comparison operators ignoring the flags, but that may be exposed through to the user in some places and I am not sure we can change it.

I guess this happens because in Verilog the string compares equal to the corresponding constant, and since Yosys is mainly used with Verilog nobody objected to this. I think that a good way to fix this issue woudl be to emit a warning (while comparing unequal) so that any such desired reliance can be found and retained with an explicit conversion.

Reply:
Note: I wrote that not having seen the prior comments (github pls). I'll respond to it more properly soon when I have enough time to understand the situation some more

Reply:
> I'll respond to it more properly soon when I have enough time to understand the situation some more

The key takeaway is this: not to upset users we need the choice of backing for Const to be independent of whether the `STRING` flag is set, which I think is already the case with this PR.

Reply:
Yep